CONTRIBUTING.md

```diff
@@ -205,6 +205,209 @@ Please follow the coding style of the project. Material-UI uses prettier and esl
+
+### Public Component
+
+refer to components that are exported from `@material-ui/core` or `@material-ui/lab`.
+
+```tsx
+import { SxProps } from '@material-ui/system';
+import {
+  experimentalStyled as styled,
+  unstable_useThemeProps as useThemeProps,
+  Theme,
+} from '@material-ui/core/styles';
+
+// 1st: props interface of the component with `classes`, `sx` and other props in alphabetical order
+export interface FooProps {
+  // add comment to each key for api docs
+  /**
+   * Override or extend the styles applied to the component.
+   */
+  classes?: {
+    /** Styles applied to the root element. */
+    root?: string;
+    /** Styles applied to the root element if `disableMargin=false`. */
+    bar?: string;
+    /** Pseudo-class applied to the root element if `disabled=true`. */
+    disabled?: string;
+  };
+  ...props
+  /**
+   * The system prop that allows defining system overrides as well as additional CSS styles.
+   */
+  sx?: SxProps<Theme>;
+}
+// in case that it extends from public component
+// export interface FooProps extends TypographyProps
+
+// 2nd: ClassKey
+export type FooClassKey = keyof NonNullable<FooProps['classes']>;
+// in case that props interface require generic argument, pass `unknown`
+// Ex. export type FooClassKey = keyof NonNullable<FooProps<unknown>['classes']>;
+
+// 3rd: classes generator & utility
+export function getFooUtilityClass(slot: string) {
+  return generateUtilityClass('MuiFooDay', slot);
+}
+// make sure it has ClassKey as arg in generic
+export const fooClasses = generateUtilityClasses<FooClassKey>('MuiFoo', [
+  'root',
+  'bar',
+  'disabled',
+]);
+
+const useUtilityClasses = (styleProps: FooProps & { extraProp: boolean}) => {
+  // extraProp might be the key/value from react context that this component access
+  const {
+    bar,
+    disabled,
+    classes,
+  } = styleProps;
+
+  const slots = {
+    root: [
+      'root',
+      bar && 'bar',
+      disabled && 'disabled',
+    ],
+  };
+
+  return composeClasses(slots, getFooUtilityClass, classes);
+};
+
+// 4th: StyledComponent, similar to JS convention
+// add additional props in between to append props interface to StyledComponent
+const FooRoot = styled(
+  Typography,
+  {},
+  {
+    name: 'MuiFoo',
+    slot: 'Root',
+    overridesResolver: (props, styles) => styles.root,
+  },
+)<{ component?: React.ElementType }>(...);
+
+// 5th: component declaration
+const Foo = React.forwardRef<HTMLSpanElement, FooProps>(function Foo(inProps, forwardedRef) => {
+  // pass args like this, otherwise will get error about theme at return section
+  const props = useThemeProps<Theme, FooProps, 'MuiFoo'>({
+    props: inProps,
+    name: 'MuiFoo',
+  });
+  const { children, className, ...other } = props
+
+  // ...implementation
+
+  const styleProps = { ...props, ...otherValue }
+
+  const classes = useUtilityClasses(styleProps);
+
+  return (
+    <FooRoot
+      ref={forwardedRef}
+      component="span" // passed to Typography
```

Comment:
```suggestion
```
Not important

Reply:
I suggest to keep it to show why the code below is needed.
```tsx
const FooRoot = styled(Typography, {}, { skipSx: true })<{ component?: React.ElementType }>({ ... })
```